text/0000-placement-traits.md

```diff
@@ -0,0 +1,312 @@
+pub trait Boxer<Data: ?Sized, A>: Sized
+    where A: Allocator
+{
+    /// The place that will negotiate the storage of the data.
+    type Place: Place<Data, Owner=Self>;
+
+    /// Creates a globally fresh place from a given allocator.
+    fn make_place(allocator: A) -> Self::Place
+        where Data: Sized;
+}
+
+/// Both `PLACE <- EXPR` and `box EXPR` desugar into expressions
+/// that allocate an intermediate "place" that holds uninitialized
+/// state.  The desugaring evaluates EXPR, and writes the result at
+/// the address returned by the `pointer` method of this trait.
+///
+/// A `Place` can be thought of as a special representation for a
+/// hypothetical `&uninit` reference (which Rust cannot currently
+/// express directly). That is, it represents a pointer to
+/// uninitialized storage.
+///
+/// The client is responsible for two steps: First, initializing the
+/// payload (it can access its address via `pointer`). Second,
+/// converting the agent to an instance of the owning pointer, via the
+/// appropriate `finalize` method.
+///
+/// If evaluating EXPR fails, then the destructor for the
+/// implementation of Place is run to clean up any intermediate state
+/// (e.g. deallocate box storage, pop a stack, etc).
+pub unsafe trait Place<Data: ?Sized> {
+    /// `Owner` is the type of the end value of both `PLACE <- EXPR` and
+    /// `box EXPR`.
+    ///
+    /// Note that when `PLACE <- EXPR` is solely used for side-effecting an
+    /// existing data-structure, e.g. `Vec::back`, then `Owner` need not carry
+    /// any information at all (e.g. it can be the unit type `()` in that
+    /// case).
+    type Owner;
+
+    /// Returns the address where the input value will be written.
+    /// Note that the data at this address is generally uninitialized,
+    /// and thus one should use `ptr::write` for initializing it.
+    fn pointer(&mut self) -> *mut Data;
+
+    /// Converts self into the final value, shifting deallocation/cleanup
+    /// responsibilities (if any remain), over to the returned instance of
+    /// `Owner` and forgetting self.
+    unsafe fn finalize(self) -> Self::Owner;
+}
+
+```
+
+First, `box` desugaring constructs the allocator with `Default::default()`. This
+means that `let boxed: Box<_, A> = box value;` works for all allocators `A:
+Default`. It's reasonable to construct new allocators on the fly like this
+because allocators are intended to be passed to collection constructors by
+value.
+
+Additionally, we define the following blanket impl that turns every `Allocator`
+into a `Placer` for all `Boxer`s.
+
+```rust
+impl<T, D, B> Placer<D, B> for T
+    where T: Allocator,
+          B: Boxer<D, T>
+{
+    type Place = B::Place;
+
+    fn make_place(self) -> Self::Place
+        where Self: Sized
+    {
+        B::make_place(self)
+    }
+}
+```
+
+This means that `let boxed_thing: Type<_> = HEAP <- thing` works out of the box
+as long as `HEAP` is an `Allocator` and `Type` implements `Boxer`.
```

Comment:
The original intention of `PLACE <- thing` was to allow placing into custom allocators. I'm probably going to change this to https://github.com/rust-lang/rfcs/pull/1426#issuecomment-167160921 which can be added after the allocator API is merged (I think).